File: lintcode/python/1409_matrix_finding_number.py
#!/usr/bin/python -t

class Solution:
    """
    @param mat: The matrix
    @return: The answer
    """
    def findingNumber(self, mat):
        # Write your code here
        n = len(mat)
        
        first_set = set(mat[0])
        
        for i in range(1, n):
            row_set = set(mat[i])
            first_set = first_set & row_set
            
        return -1 if len(first_set) == 0 else min([i for i in first_set])
        

class Solution:
    """
    @param mat: The matrix
    @return: The answer
    用map按照题意模拟即可
    """
    def findingNumber(self, mat):
        # Write your code here
        hashSet = [0 for i in range(100000 + 1)]
        n = len(mat)
        for mati in mat:
            vis = {}
            for x in mati:
                vis[x] = 1
            for key in vis:
                hashSet[key] += 1
        ans = -1
        for i in range(1, 100000 + 1):
            if hashSet[i] == n:
                ans = i
                break
        return ans

File: lintcode/golang/1409_matrix_finding_number.go
/**
 * @param mat: The matrix
 * @return: The answer
 */
func findingNumber (mat [][]int) int {
    // Write your code here
    m := len(mat)
    if m == 0 {
        return -1
    }

    n := len(mat[0])

    d := make(map[int]bool)

    for i := 0; i < m; i++ {
        level := make(map[int]bool)
        for j := 0; j < n; j++ {
            if i == 0 {
                d[mat[i][j]] = true
            }

            if d[mat[i][j]] == true {
                level[mat[i][j]] = true
            }
        }
        d = level
    }

    if len(d) != 0 {
        ret := 0
        for k, _ := range d {
            if k < ret || ret == 0 {
                ret = k
            }
        }
        return ret
    }

    return -1
}

